### PR TITLE
Issue #2470 - Enhance analysts ability to update "Action needed" email text [nl]

### LINKED ISSUE #2470: Enhance analysts ability to update "Action needed" email text
### Issue description

In #1901, the analyst can select a specific action needed and is provided with an in-line email template to modify before sending out.  In this issue, we will address how the system will handle when the analyst decides to update the email content of the previously sent out email. Currently, if the analyst edits the email content and save, it will send another email. We want to add more safeguards for the analyst to confirm their action.  

### Acceptance criteria
Detailed AC below with screenshots. Refer to [Figma](https://www.figma.com/design/ghfsk20NAtP1xyjsFOpxHg/Django-Admin-App-Wireframes?node-id=36-839&t=JleCUDG1OzESQ0G6-11) for more detail if needed. Note: the Figma is not totally up to date with styling of django admin so use what we have in production over what's in Figma. 
- [x] Change field label to email
- [x] If no action needed reason is selected, display text in the empty Email field "Select an action needed reason to see email" 
Screenshot 2024-07-17 at 12 49 41 PM
- [x] When reason is selected, display read-only view of email Image
- [x] When user clicks "Edit email" button, field changes to editable (this only happens the first time a action needed reason is selected) If an email has already been sent, behavior changes (see AC below) 

Screenshot 2024-07-17 at 12 54 53 PM

- [x] If an email has been sent already and the edit email button is clicked, show warning modal
Screenshot 2024-07-17 at 12 57 34 PM

- [x] If user clicks "Yes, continue editing" the modal closes and the field turns editable. If they click "Cancel" the email reverts back to its state before clicking the edit button.

- [ ] If user Saves the request after action needed reason is selected, the read-only view of the email changes. Heading changes to: "Email sent to the creator" and a new green checkmark appears. Also, helper text below changes to say: "This email has been sent to the creator of this request" 
Image

### Additional context

_No response_

### Links to other issues

Blocked by:
- #1901 

Blocks:
- #2355

### PR DESCRIPTION
**Pull request by CocoByte**
## Ticket 2470

Resolves #2470

## Changes

- Updated behavior of "Action needed" email text in the admin section. (see ticket ACs)

## Code Review Verification Steps
Code is deployed to [sandbox-nl](https://getgov-nl.app.cloud.gov/)

Go to /admin, open a domain request, change status to "Action Needed" and verify the email section behaves as specified in the ticket ACs (it is detailed out in the ticket body)

#### Satisfied acceptance criteria and met development standards

- [ ] Met the acceptance criteria, or will meet them in a subsequent PR
- [ ] Created/modified automated tests
- [ ] Added at least 2 developers as PR reviewers (only 1 will need to approve)
- [ ] Messaged on Slack or in standup to notify the team that a PR is ready for review
- [ ] Changes to “how we do things” are documented in READMEs and or onboarding guide
- [ ] If any model was updated to modify/add/delete columns, makemigrations was ran and the associated migrations file has been commited.

#### Ensured code standards are met (Original Developer)

- [ ] All new functions and methods are commented using plain language
- [ ] Did dependency updates in Pipfile also get changed in requirements.txt?
- [ ] Interactions with external systems are wrapped in try/except
- [ ] Error handling exists for unusual or missing values

#### Validated user-facing changes (if applicable)

- [ ] New pages have been added to .pa11yci file so that they will be tested with our automated accessibility testing
- [ ] Checked keyboard navigability
- [ ] Tested general usability, landmarks, page header structure, and links with a screen reader (such as Voiceover or ANDI)
- [ ] Add at least 1 designer as PR reviewer

#### Reviewed, tested, and left feedback about the changes

- [ ] Pulled this branch locally and tested it
- [ ] Reviewed this code and left comments
- [ ] Checked that all code is adequately covered by tests
- [ ] Made it clear which comments need to be addressed before this work is merged
- [ ] If any model was updated to modify/add/delete columns, makemigrations was ran and the associated migrations file has been commited.

#### Ensured code standards are met (Code reviewer)

- [ ] All new functions and methods are commented using plain language
- [ ] Interactions with external systems are wrapped in try/except
- [ ] Error handling exists for unusual or missing values
- [ ] (Rarely needed) Did dependency updates in Pipfile also get changed in requirements.txt?

#### Validated user-facing changes as a developer

- [ ] New pages have been added to .pa11yci file so that they will be tested with our automated accessibility testing
- [ ] Checked keyboard navigability
- [ ] Meets all designs and user flows provided by design/product
- [ ] Tested general usability, landmarks, page header structure, and links with a screen reader (such as Voiceover or ANDI)

- [ ] Tested with multiple browsers, the suggestion is to use ones that the developer didn't (check off which ones were used)
  - [ ] Chrome
  - [ ] Microsoft Edge
  - [ ] FireFox
  - [ ] Safari

- [ ] (Rarely needed) Tested as both an analyst and applicant user

**Note:** Multiple code reviewers can share the checklists above, a second reviewers should not make a duplicate checklist

#### Verified that the changes match the design intention

- [ ] Checked that the design translated visually
- [ ] Checked behavior
- [ ] Checked different states (empty, one, some, error)
- [ ] Checked for landmarks, page heading structure, and links
- [ ] Tried to break the intended flow

#### Validated user-facing changes as a designer

- [ ] Checked keyboard navigability
- [ ] Tested general usability, landmarks, page header structure, and links with a screen reader (such as Voiceover or ANDI)

- [ ] Tested with multiple browsers (check off which ones were used)
  - [ ] Chrome
  - [ ] Microsoft Edge
  - [ ] FireFox
  - [ ] Safari

- [ ] (Rarely needed) Tested as both an analyst and applicant user

### CODE REVIEW

**Comment by gabydisarli** [comment]
Hi @CocoByte  I'll be doing the design review

File: src/registrar/assets/js/get-gov-admin.js [hunk 1]
```diff
@@ -542,17 +553,41 @@ function initializeWidgetOnList(list, parentId) {
...
+            showEmail(true)
+        });
+
+        // Event delegation for data-close-modal buttons 
+        // (since manually opening the modal interferes with how this normally works)
+        document.addEventListener('click', function(event) {
+            if (event.target.matches('[data-close-modal]')) {
+                setModalVisibility(actionNeededEmailAlreadySentModal, false)
+            }
         });
```
[the remark this answers is not in the document]
Comment (CocoByte) [reply src/registrar/assets/js/get-gov-admin.js:583]: I saw that.  Actually, I think this way is a bit more straight-forward and leaves the DOM cleaner.  It also seems to be easier to follow and provides an easily extensible way for handling custom behavior.  There are trade-offs for both, for sure, but for now I think I will leave it as-is.

File: src/registrar/assets/js/get-gov-admin.js [hunk 2]
```diff
@@ -562,35 +597,93 @@ function initializeWidgetOnList(list, parentId) {
...
+        hideElement(actionNeededEmail.parentElement)
+
+        if (reason) {
+            if (reason === "other") {
+                // Hide email preview and show this text instead
+                showPlaceholderText("No email will be sent");
             }
+            else {
+                // Always show readonly view of email to start
+                showEmail(false)
```
[the remark this answers is not in the document]
Comment (CocoByte) [reply src/registrar/assets/js/get-gov-admin.js:644]: Really good suggestion! I like it.

**Comment by gabydisarli** [comment]
- Can we change the helper text: `Select an action needed reason to see email` to the same font/style that we use for helper text across django admin?
- Please change `E-mail` to `Email` for the label

**Comment by gabydisarli** [comment]
I'm seeing some odd things with the field like it's not responsive? This is on Chrome just on my macbook pro.
Screenshot 2024-08-19 at 3 12 12 PM

**Comment by CocoByte** [comment]
> I'm seeing some odd things with the field like it's not responsive? This is on Chrome just on my macbook pro. Screenshot 2024-08-19 at 3 12 12 PM

This is a CSS issue with the sandbox.  I'll look into it and try to fix it

File: src/registrar/templates/django/admin/includes/detail_table_fieldset.html [hunk 3]
```diff
@@ -145,20 +145,102 @@
...
+                        <div class="usa-modal__main">
+                            <h2 class="usa-modal__heading" id="modal-1-heading">
+                                Are you sure you want to edit this email?
+                            </h2>
+                            <div class="usa-prose">
+                                <p>
+                                    The creator of this request already received an email for this status/reason: 
+                                </p>
+                                <ul>
+                                    <li class="font-body-sm">Status: <b>Action needed</b></li>
```
[the remark this answers is not in the document]
Comment (CocoByte) [reply src/registrar/templates/django/admin/includes/detail_table_fieldset.html:170]: Yeah, I sensed a refactor opportunity, but sometimes trying to "refactor early" causes more issues than it solves, so I went for simpler-is-better in the meantime.

**Comment by gabydisarli** [comment]
Hi @CocoByte !
- There must be a bug. I edit the email, click save and edit, and it reverts back to the original unedited version. I do get the modal warning me that the email will be sent. When I edit the email and click "Save" I do see it change really quick to the sent version. But when I navigate back to the request, it's reverted back to the original state.

Some other things
- Left-align helper text "This email will be sent to the creator of this request after saving" right now it's centered under the field
- Change text inside of email field (like the actual generated email text) to match the other body text of the fields. I'm seeing 0.8125rem on inspect
- Not sure why, but I don't see the little grabby handle in the bottom right corner. Im not seeing it in the notes field as well so it must be something else not related to this PR. Still, can we put in some restrictions on how the field can be expanded? For example, on the notes field, I can drag it down and to the right, but it will not stretch beyond the page. The email field can be pulled off the page.

**Comment by gabydisarli** [comment]
I'm still seeing some odd behavior. When I don't do anything to the auto generated email and hit "Save and continue" button, I do see the Email sent to submitter version, but the email in the field has weird styling. If I do click edit email and change the text, then hit the "Save and continue" button, the text update persists but the field goes back to the version that says "Auto-generated email that will be sent to the creator"
Screenshot 2024-08-20 at 9 45 03 PM
Screenshot 2024-08-20 at 9 48 12 PM

These are some previous things from a previous comment I'm still seeing:

- Change text inside of email field (like the actual generated email text) to match the other body text of the fields. I'm seeing 0.8125rem on inspect
- Can we put in some restrictions on how the field can be expanded? For example, on the notes field, I can drag it down and to the right, but it will not stretch beyond the page. The email field can be pulled off the page (no max width).

**Comment by gabydisarli** [comment]
Hi @rachidatecs a few more things:

- make checkbox icon green (5E9732)
- change hatch marks to quotation marks on "Save" and "Cancel" in the modal

Accessibility findings:
- when the modal is triggered it doesn't trap the keyboard tabs so it's not navigable by keyboard only
- On the email field I'm getting this error from ANDI: "danger[Textarea has no accessible name, associated <label>, or [title].](https://www.ssa.gov/accessibility/andi/help/alerts.html?no_name_form_element)"

cc @AnnaGingle

**Comment by AnnaGingle** [comment]
Taking a look 👀

**Comment by AnnaGingle** [comment]
@gabydisarli is correct. The ANDI output for the (noneditable) Email: field should be `Email:`.

Additionally, when the `Email:` field is in edit mode the ANDI output is `Email:` but it should be `Email: | This email will be sent to the creator of this request after saving`.

**Comment by rachidatecs** [comment]
- make checkbox icon green (5E9732): DONE, but I made it the same green we have on our checkmarks on request sidenav (text-green, #538200)
- change hatch marks to quotation marks on "Save" and "Cancel" in the modal: DONE
- When viewing at smaller screen sizes, the field lockup doesn't respond well and "breaks": NOT DONE, this is probably a big deal thing
- When the modal is triggered it doesn't trap the keyboard tabs so it's not navigable by keyboard only: DONE
On the email field I'm getting this error from ANDI: "danger[Textarea has no accessible name, associated , or [title].](https://www.ssa.gov/accessibility/andi/help/alerts.html?no_name_form_element)": PARTIALLY DONE (just added a name with 'Email' for now)

cc @gabydisarli @AnnaGingle 
I think that getting the last bits in would require a change in how this is implemented. We'll discuss on Monday when Nicolle's back.

**Review by gabydisarli — APPROVED** [review]
Approved!

Noting there will be a new ticket to address Anna's accessibility findings and improve the functionality of the nested field.

**Pull request merged**